### DISCUSSION TITLE
Testing: runner.invoke reuses 'app' object, causing cached properties to exist when they're not supposed to

### DISCUSSION THREAD

**Discussion by Dymstro** [fastapi/typer#1318] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typer.testing import CliRunner
from .main import app

runner = CliRunner()

def test_list_databases() -> None:
    # This command causes an internal 'databases' object to be cached.

    result = runner.invoke(app, ["databases", "list"]
    assert result.exit_code == 0


def test_create_database() -> None:
    # This command accesses some cached properties (using cached-property package).
    # Those objects aren't supposed to exist yet, but they were already created in previous tests.
    # This causes the command to use old data and crash.

    result = runner.invoke(app, ["databases", "create", "tests-db", "MariaDB"])
    assert result.exit_code == 0
```

### Description

CliRunner.invoke reuses 'app' causing different behavior than just running them outside tests.

My code uses some cached properties that get cached in earlier tests.
This causes some commands to use old data, which causes it to crash.
Normally this would never happen as the program exits after a single command.

### Operating System

Linux

### Typer Version

0.4.0

### Python Version

3.7

**Comment by DvdNss** [comment]
happening for me as well, some options get cached and re-used in the following tests, making the tests fail

**Comment by YuriiMotov** [comment, marked as answer]
This is not a Typer's issue..

You can solve this by re-importing the app on every test:

```py
import helloworld.main

...
def test_list_databases():
    importlib.reload(helloworld.main)
    result = runner.invoke(helloworld.main.app, ...)
    ...
```

<details><summary>Working code example</summary>

**helloworld/main.py**
```py
from functools import cached_property

import typer

class DatabaseManager:
    @cached_property
    def databases(self) -> list[str]:
        print(">> Creating databases list")
        return ["default-db"]

    def create(self, name: str) -> None:
        print(f">> Creating database {name}")
        self.databases.append(name)

manager = DatabaseManager()
app = typer.Typer()

@app.command()
def list_databases():
    for db in manager.databases:
        typer.echo(db)

@app.command()
def create_database(name: str):
    manager.create(name)
    typer.echo(f"Database {name} created")
```

**helloworld/test.py**
```py
import importlib

from typer.testing import CliRunner

import helloworld.main

runner = CliRunner()

def test_list_databases():
    importlib.reload(helloworld.main)
    result = runner.invoke(helloworld.main.app, ["list-databases"])
    assert result.exit_code == 0
    assert ">> Creating databases list" in result.stdout


def test_create_database():
    importlib.reload(helloworld.main)
    result = runner.invoke(helloworld.main.app, ["create-database", "test-db"])
    assert result.exit_code == 0
    assert ">> Creating databases list" in result.stdout

```

</details>

**Reply by WilliamDEdwards** [reply]
> This is not a Typer's issue..

I agree that with the current implementation, i.e. `app` being a singleton, this is expected behaviour.

However, seeing as how the lifecycle of a CLI app is usually short, and tests should be isolated from each other, I would argue this is more of an architectural gray area.
